### PR TITLE
Feature/harmony

### PR DESCRIPTION
**Pull request by tokejepsen**
This is the code from the PR awaiting merge here; https://github.com/getavalon/core/pull/538

Pype core seems to be missing a PR from Avalon core; https://github.com/getavalon/core/pull/529, which it why you can see commit from David.
It should be fine to merge, but better check the functionality of the merged PR.

### CODE REVIEW

File: avalon/harmony/TB_sceneOpened.js [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+  self.process_request = function(request)
+  {
+    self.log_debug("Processing: " + JSON.stringify(request));
+    var result = null;
+
+    if (request["function"] != null)
+    {
+      try
+      {
+        var func = eval(request["function"]);
```
[the remark this answers is not in the document]
Comment (tokejepsen) [reply avalon/harmony/TB_sceneOpened.js:43]: This has been commented on in the main PR, but this eval is necessary to evaluate string code coming into Harmony.

**Comment by mkolar** [comment]
@iLLiCiTiT can you please have a look at this because of the GUI updates from David? I have a feeling they might have to be ported to us carefully.

File: avalon/harmony/TB_sceneOpened.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (iLLiCiTiT) [reply avalon/harmony/TB_sceneOpened.js:43]: I would prefer to not send function (code as string) but rather have implemented those functions in javascript and send only instruction which function will be used in request.

**Comment by iLLiCiTiT** [comment]
> @iLLiCiTiT can you please have a look at this because of the GUI updates from David? I have a feeling they might have to be ported to us carefully.

I don't like the way how it's approached but that would be neverending story, so yes it won't break anything :)

But possible problems I ran into when I went through the code is that communication server is running in thread of main process, which is pype I expect, and uses io.Session and environments expecting nothing else in the process will change it's values... So only one host at the time and hopefully none of pype's module will change those values... Or I missed something @tokejepsen ?

File: avalon/harmony/TB_sceneOpened.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tokejepsen) [reply avalon/harmony/TB_sceneOpened.js:43]: The issue with having the function implemented in javascript is that any code we havent thought of, will need to be implemented in the core.
By being able to send code and functions we can do all the work in plugins and keep the core intact.

**Comment by tokejepsen** [comment]
> So only one host at the time and hopefully none of pype's module will change those values... Or I missed something @tokejepsen ?

Not entirely sure I'm getting the full picture here. You can change context if thats what you mean?

File: avalon/harmony/TB_sceneOpened.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (iLLiCiTiT) [reply avalon/harmony/TB_sceneOpened.js:43]: I don't like using eval, but I see the issue and don't know about any easy solution for that.

**Comment by iLLiCiTiT** [comment]
> Not entirely sure I'm getting the full picture here. You can change context if thats what you mean?

I apologize for my explanation skills. Long story short: When and where(which process) is launched the socket server?

**Comment by tokejepsen** [comment]
> I apologize for my explanation skills. Long story short: When and where(which process) is launched the socket server?

The socket server is launched in a thread of the main python process. The main python process is our launcher here; https://github.com/pypeclub/pype-config/pull/25/files#diff-088ac42f7e32e9c191247eea0528907d

I might have this wrong, but were you thinking that the python process was main pype process?

**Comment by iLLiCiTiT** [comment]
> I might have this wrong, but were you thinking that the python process was main pype process?

Yes, exactly and now I see :)

Good for me.

**Pull request merged**